### PR TITLE
Send output values to transforms for dependency tracking

### PR DESCRIPTION
**Pull request by Frassle**
# Description

This engine change is necessary for correct dependency tracking of properties through transform functions.

Unlike other parts of the runtime/provider/engine interface transform functions do not have a property dependencies map sent or returned. Transforms rely entirely on the dependency arrays in output values for their dependency tracking.

This change takes the property dependency map and upgrades all the input properties to be output values tracking those same dependencies (if a property doesn't have any dependencies it doesn't need to be upgraded to an output value).
This upgrade is only done if we're using transform functions, there's currently no need to do this unless we're using transforms.

After running the transforms we downgrade the output values back to plain/secret/computed values if we're registering a custom resource. If we're registering a component resource we just leave all the properties upgraded as output values.
We also rebuild the resources dependencies slice and propertyDependency map with the dependencies recorded in the transformed properties. If the transform didn't change the properties this will just be the same data we encoded into the output values in the properties structure.

There are a couple of things to keep in mind with this change. 

Firstly using a transform can cause a property that was being sent to a component provider as just a `resource.NumberProperty` to start being sent as an `OutputProperty` with the `NumberProperty` as its element. Component providers _should_ handle that, but it's feasible that it could break some component code. This is a fairly limited blast radius as it will only happen when that user starts using a transform function.

Secondly, we can't know in the engine if a property dependencies are from a top level output value or a nested output. That is SDKs send both of the following property shapes the same way to the engine for custom resources:
```
A) Output(Array([Number(1)]), ["dep"])
---
B) Array([Output(Number(1), ["dep"])
```
Currently both would get upgraded to `Output(Array([Number(1)]), ["dep"])`. For a component resource the SDK will send output values, and as long as they define a superset of the dependencies in the property map then we don't add the top-level output tracking the same dependency set.

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: pkg/resource/deploy/source_eval.go [hunk 1]
```diff
@@ -2147,6 +2180,35 @@ func downgradeOutputValues(v resource.PropertyMap) resource.PropertyMap {
...
+func upgradeOutputValues(
+	v resource.PropertyMap, propertyDependencies map[resource.PropertyKey][]resource.URN,
+) resource.PropertyMap {
+	// We assume that by the time this is being called we've upgraded all Secret/Computed values to outputs. We just
+	// need to add the dependency information from propertyDependencies.
+
+	result := make(resource.PropertyMap)
+	for k, pv := range v {
+		if deps, has := propertyDependencies[k]; has {
+			var output resource.Output
```
Comment (Frassle) [inline pkg/resource/deploy/source_eval.go:2192]: This isn't _quite_ right, we should see if `pv` already defines the same dependencies as `deps` (because it's already an output value, or has nested output values)

**Comment by Frassle** [comment]
This needs a bit more set handling for the dependency arrays, going to try and get https://github.com/pulumi/pulumi/pull/15671 and maybe some other small changes in first to minimise diff size.

File: pkg/resource/deploy/source_eval.go [hunk 2]
```diff
@@ -1605,13 +1644,42 @@ func (rm *resmon) RegisterResource(ctx context.Context,
...
+			addOutputDependencies(deps, output)
+			propertyDependencies[key] = deps
+
+			// Also add these to the overall dependencies
+			dependencies = dependencies.Union(deps)
+		}
+	} else {
+		// If we ran transforms we would have merged all the dependencies togther already, but if we didn't we want to
+		// ensure any output values add their dependencies to the dependencies map we send to the provider.
+		for key, output := range props {
```
Comment (tgummerer) [inline pkg/resource/deploy/source_eval.go:1671]: I'm not sure I'm following this change properly.  Since we didn't upgrade the props in the case where there are no transforms, couldn't we just use the old code from https://github.com/pulumi/pulumi/pull/15637/files#diff-984b3279d61b588030474831898bb76d093d863ca2d0015bb081d842602a9ab7L1665?  Or why is this change necessary?

**Review by tgummerer — COMMENTED** [review]
This looks good go me, with one question inline.

**Comment by justinvp** [comment]
> Firstly using a transform can cause a property that was being sent to a component provider as just a `resource.NumberProperty` to start being sent as an `OutputProperty` with the `NumberProperty` as its element.

Does this only happen if that property has dependencies?

File: pkg/resource/deploy/source_eval.go (hunk 2 above)
Comment (Frassle) [reply pkg/resource/deploy/source_eval.go:1671]: This is pretty much that code but for Set instead of slice.

**Review by tgummerer — APPROVED** [review]
